client/go/outline/iptable/packet_listener.go

```diff
@@ -0,0 +1,319 @@
+// packets over various underlying [net.PacketConn]s based on IP addresses.
+// It is returned by [PacketListener.ListenPacket].
+//
+// For outgoing packets (WriteTo), it determines the appropriate underlying
+// connection using an [IPTable] and a default listener. New underlying
+// connections are established on-demand.
+// For incoming packets (ReadFrom), it aggregates packets received from all
+// active underlying connections (both specific and default) into a single channel.
+type packetConn struct {
+	defaultListener transport.PacketListener
+	defaultConn     net.PacketConn
+
+	listenerTable IPTable[transport.PacketListener]
+	connMap       map[netip.Addr]net.PacketConn
+	connMapLock   sync.Mutex
+
+	forwardingContext            context.Context
+	closePacketForwardingContext context.CancelFunc
+	forwardedPackets             chan incomingPacket
+	forwardCounter               sync.WaitGroup
+
+	lastRemainingPacket     incomingPacket
+	lastRemainingPacketLock sync.Mutex
+}
+
+var _ net.PacketConn = (*packetConn)(nil)
+
+func newPacketConn(
+	parentContext context.Context,
+	listenerTable IPTable[transport.PacketListener],
+	defaultListener transport.PacketListener,
+) (*packetConn, error) {
+	forwardingContext, closeForwardingContext := context.WithCancel(parentContext)
+
+	var defaultConn net.PacketConn
+	if defaultListener != nil {
+		connAttempt, err := defaultListener.ListenPacket(forwardingContext)
+
+		if err != nil {
+			defaultConn = nil
+		} else {
+			defaultConn = connAttempt
+		}
+	}
+
+	conn := &packetConn{
+		closePacketForwardingContext: closeForwardingContext,
+		connMap:                      make(map[netip.Addr]net.PacketConn),
+		connMapLock:                  sync.Mutex{},
+		defaultConn:                  defaultConn,
+		defaultListener:              defaultListener,
+		forwardedPackets:             make(chan incomingPacket, packetQueueSize),
+		forwardingContext:            forwardingContext,
+		listenerTable:                listenerTable,
+		lastRemainingPacket:          incomingPacket{},
+		lastRemainingPacketLock:      sync.Mutex{},
+	}
+
+	if conn.defaultConn != nil {
+		conn.forwardPackets(conn.defaultConn)
+	}
+
+	return conn, nil
+}
+
+func (conn *packetConn) ReadFrom(result []byte) (n int, addr net.Addr, err error) {
+	if len(conn.lastRemainingPacket.data) > 0 {
+		conn.lastRemainingPacketLock.Lock()
+		numBytes := copy(result, conn.lastRemainingPacket.data)
+		conn.lastRemainingPacket.data = conn.lastRemainingPacket.data[numBytes:]
+		conn.lastRemainingPacketLock.Unlock()
+
+		return numBytes, conn.lastRemainingPacket.addr, nil
+	}
+
+	packet, ok := <-conn.forwardedPackets
+	if !ok {
+		return 0, nil, net.ErrClosed
+	}
+
+	numBytes := copy(result, packet.data)
+
+	conn.lastRemainingPacketLock.Lock()
+	conn.lastRemainingPacket.data = packet.data[numBytes:]
+	conn.lastRemainingPacket.addr = packet.addr
+	conn.lastRemainingPacketLock.Unlock()
+
+	return numBytes, packet.addr, nil
```

Comment:
There are multiple race conditions here, it's hard to make it right.

Actually it's safe to discard UDP packets, my bad (I was thinking about TCP before). Let's remove `lastRemainingPacket`. Sorry about that.

```suggestion
    packet, ok := <-conn.forwardedPackets
    if !ok {
        return 0, nil, net.ErrClosed
    }

    // Copy what fits and discard the rest according to UDP standard
    n = copy(result, packet.data)
    return n, packet.addr, nil
```